### PR TITLE
[EraVM] Enable AliasAnalysis for all location sizes

### PR DESCRIPTION
**Pull request by vladimirradosavljevic**

### CODE REVIEW

**Comment by zksync-admin-bot2** [comment]
Benchmark results:
```
╔═╡ Size (-%) ╞════════════════╡ All M3B3 ╞═╗
║ Mean                                0.059 ║
║ Best                               68.088 ║
║ Worst                              -4.247 ║
║ Total                               0.165 ║
╠═╡ Cycles (-%) ╞══════════════╡ All M3B3 ╞═╣
║ Mean                                0.032 ║
║ Best                               53.529 ║
║ Worst                              -2.619 ║
║ Total                               0.006 ║
╠═╡ Ergs (-%) ╞════════════════╡ All M3B3 ╞═╣
║ Mean                                0.012 ║
║ Best                               39.258 ║
║ Worst                              -1.765 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞════════════════╡ All MzB3 ╞═╗
║ Mean                                0.034 ║
║ Best                               28.054 ║
║ Worst                              -0.784 ║
║ Total                               0.054 ║
╠═╡ Cycles (-%) ╞══════════════╡ All MzB3 ╞═╣
║ Mean                                0.018 ║
║ Best                               39.594 ║
║ Worst                             -16.364 ║
║ Total                               0.003 ║
╠═╡ Ergs (-%) ╞════════════════╡ All MzB3 ╞═╣
║ Mean                                0.008 ║
║ Best                               18.824 ║
║ Worst                              -0.962 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞══════════╡ Real life M3B3 ╞═╗
║ Mean                               -0.003 ║
║ Best                                1.290 ║
║ Worst                              -4.052 ║
║ Total                               0.036 ║
╠═╡ Cycles (-%) ╞════════╡ Real life M3B3 ╞═╣
║ Mean                                0.001 ║
║ Best                                0.095 ║
║ Worst                               0.000 ║
║ Total                               0.001 ║
╠═╡ Ergs (-%) ╞══════════╡ Real life M3B3 ╞═╣
║ Mean                                0.001 ║
║ Best                                0.077 ║
║ Worst                              -0.008 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞══════════╡ Real life MzB3 ╞═╗
║ Mean                                0.014 ║
║ Best                                1.003 ║
║ Worst                              -0.150 ║
║ Total                               0.012 ║
╠═╡ Cycles (-%) ╞════════╡ Real life MzB3 ╞═╣
║ Mean                                0.002 ║
║ Best                                0.595 ║
║ Worst                              -0.712 ║
║ Total                               0.003 ║
╠═╡ Ergs (-%) ╞══════════╡ Real life MzB3 ╞═╣
║ Mean                                0.002 ║
║ Best                                0.262 ║
║ Worst                              -0.003 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

```

File: llvm/lib/Target/EraVM/EraVMAliasAnalysis.cpp [hunk 1]
```diff
@@ -46,14 +46,10 @@ void EraVMAAWrapperPass::getAnalysisUsage(AnalysisUsage &AU) const {
   AU.setPreservesAll();
 }
 
-static Optional<APInt> getConstStartLoc(const MemoryLocation &Loc) {
-  // TODO: CPR-1337 Enable all pointer sizes.
-  constexpr unsigned ByteWidth = 32;
-  if (Loc.Size != ByteWidth)
-    return None;
-
+static Optional<APInt> getConstStartLoc(const MemoryLocation &Loc,
```
Comment (akiramenai) [inline llvm/lib/Target/EraVM/EraVMAliasAnalysis.cpp]: Haven't noticed in the first reading. `std::optional`, afaik LLVM is going to deprecate their Optional, Variant.

**Review by akiramenai — APPROVED** [review]
I wonder, have we expected performance regressions from this patch? I mean not overall, but for particular tests? It feels to me that the analysis became more precise, so we might want to create a follow up investigation ticket to address negative changes if unexpected. Imo low to mid priority is ok for it as it doesn't look that significant, but as you started to dig into the optimiser you might want to prioritize.

**Comment by akiramenai** [comment]
And please squash.

File: llvm/lib/Target/EraVM/EraVMAliasAnalysis.cpp (hunk 1 above)
Comment (vladimirradosavljevic) [reply llvm/lib/Target/EraVM/EraVMAliasAnalysis.cpp]: Changed to `std::optional`. Thanks for noticing this!

**Comment by zksync-admin-bot2** [comment]
Benchmark results:
```
╔═╡ Size (-%) ╞════════════════╡ All M3B3 ╞═╗
║ Mean                                0.323 ║
║ Best                               68.156 ║
║ Worst                             -19.835 ║
║ Total                               0.697 ║
╠═╡ Cycles (-%) ╞══════════════╡ All M3B3 ╞═╣
║ Mean                                0.198 ║
║ Best                               85.302 ║
║ Worst                             -24.390 ║
║ Total                               0.024 ║
╠═╡ Ergs (-%) ╞════════════════╡ All M3B3 ╞═╣
║ Mean                                0.131 ║
║ Best                               93.737 ║
║ Worst                             -17.143 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞════════════════╡ All MzB3 ╞═╗
║ Mean                                0.133 ║
║ Best                               45.669 ║
║ Worst                             -10.628 ║
║ Total                               0.224 ║
╠═╡ Cycles (-%) ╞══════════════╡ All MzB3 ╞═╣
║ Mean                                0.063 ║
║ Best                               62.829 ║
║ Worst                             -24.896 ║
║ Total                               0.007 ║
╠═╡ Ergs (-%) ╞════════════════╡ All MzB3 ╞═╣
║ Mean                                0.034 ║
║ Best                               79.885 ║
║ Worst                             -13.183 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞══════════╡ Real life M3B3 ╞═╗
║ Mean                                0.165 ║
║ Best                               22.172 ║
║ Worst                              -3.683 ║
║ Total                               0.163 ║
╠═╡ Cycles (-%) ╞════════╡ Real life M3B3 ╞═╣
║ Mean                                0.002 ║
║ Best                                0.103 ║
║ Worst                              -0.058 ║
║ Total                               0.022 ║
╠═╡ Ergs (-%) ╞══════════╡ Real life M3B3 ╞═╣
║ Mean                                0.001 ║
║ Best                                0.085 ║
║ Worst                              -0.007 ║
║ Total                               0.001 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞══════════╡ Real life MzB3 ╞═╗
║ Mean                                0.151 ║
║ Best                               22.123 ║
║ Worst                              -2.198 ║
║ Total                               0.165 ║
╠═╡ Cycles (-%) ╞════════╡ Real life MzB3 ╞═╣
║ Mean                               -0.001 ║
║ Best                                0.578 ║
║ Worst                              -0.718 ║
║ Total                               0.016 ║
╠═╡ Ergs (-%) ╞══════════╡ Real life MzB3 ╞═╣
║ Mean                                0.001 ║
║ Best                                0.265 ║
║ Worst                              -0.114 ║
║ Total                               0.001 ║
╚═══════════════════════════════════════════╝

```

**Comment by vladimirradosavljevic** [comment]
> I wonder, have we expected performance regressions from this patch? I mean not overall, but for particular tests? It feels to me that the analysis became more precise, so we might want to create a follow up investigation ticket to address negative changes if unexpected. Imo low to mid priority is ok for it as it doesn't look that significant, but as you started to dig into the optimiser you might want to prioritize.

Thank you for pointing this out! I didn't expect such degradations in some tests. I suggest to go through the regressions and to understand them or to fix them. I started looking into them, because ideally, they can open more opportunities for performance gain.

**Comment by zksync-admin-bot2** [comment]
Benchmark results:
```
╔═╡ Size (-%) ╞════════════════╡ All M3B3 ╞═╗
║ Mean                                0.323 ║
║ Best                               68.156 ║
║ Worst                             -19.835 ║
║ Total                               0.697 ║
╠═╡ Cycles (-%) ╞══════════════╡ All M3B3 ╞═╣
║ Mean                                0.198 ║
║ Best                               85.302 ║
║ Worst                             -24.390 ║
║ Total                               0.024 ║
╠═╡ Ergs (-%) ╞════════════════╡ All M3B3 ╞═╣
║ Mean                                0.131 ║
║ Best                               93.737 ║
║ Worst                             -17.143 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞════════════════╡ All MzB3 ╞═╗
║ Mean                                0.133 ║
║ Best                               45.669 ║
║ Worst                             -10.628 ║
║ Total                               0.224 ║
╠═╡ Cycles (-%) ╞══════════════╡ All MzB3 ╞═╣
║ Mean                                0.063 ║
║ Best                               62.829 ║
║ Worst                             -24.896 ║
║ Total                               0.007 ║
╠═╡ Ergs (-%) ╞════════════════╡ All MzB3 ╞═╣
║ Mean                                0.034 ║
║ Best                               79.885 ║
║ Worst                             -13.183 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞══════════╡ Real life M3B3 ╞═╗
║ Mean                                0.165 ║
║ Best                               22.172 ║
║ Worst                              -3.683 ║
║ Total                               0.163 ║
╠═╡ Cycles (-%) ╞════════╡ Real life M3B3 ╞═╣
║ Mean                                0.002 ║
║ Best                                0.103 ║
║ Worst                              -0.058 ║
║ Total                               0.022 ║
╠═╡ Ergs (-%) ╞══════════╡ Real life M3B3 ╞═╣
║ Mean                                0.001 ║
║ Best                                0.085 ║
║ Worst                              -0.007 ║
║ Total                               0.001 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞══════════╡ Real life MzB3 ╞═╗
║ Mean                                0.151 ║
║ Best                               22.123 ║
║ Worst                              -2.198 ║
║ Total                               0.165 ║
╠═╡ Cycles (-%) ╞════════╡ Real life MzB3 ╞═╣
║ Mean                               -0.001 ║
║ Best                                0.578 ║
║ Worst                              -0.718 ║
║ Total                               0.016 ║
╠═╡ Ergs (-%) ╞══════════╡ Real life MzB3 ╞═╣
║ Mean                                0.001 ║
║ Best                                0.265 ║
║ Worst                              -0.114 ║
║ Total                               0.001 ║
╚═══════════════════════════════════════════╝

```

**Pull request merged**